controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -52,78 +81,315 @@ boolean lock(Integer permit, long timeout, TimeUnit unit) throws InterruptedExce
+            return state != 0L && getLockMode(state) == EXCLUSIVE;
+        }
+
+        private int getPermitHolder(final long value) {
+            return (int) (value & PERMIT_MASK);
+        }
+
+        private short getCount(final long value) {
+            return (short) ((value >> COUNT_SHIFT) & SHORT_MASK);
+        }
+
+        private short getLockMode(final long value) {
+            return (short) ((value >> EXCLUSIVE_SHIFT) & SHORT_MASK);
+        }
+
+        // store creates the long state value, storing it as (short)lockMode(short)lockCount(int)permitHolder
+        // int the case of shared mode, permitHolder will be OL.
+        private long makeState(final int permit, final int count, final short lockMode) {
+            assert lockMode == EXCLUSIVE || lockMode == SHARED;
+            assert count > 0;
+            if (lockMode == SHARED && permit != 0) {
+                throw new IllegalStateException("Permit for shared mode lock must be 0.");
+            }
+            if (count < 0 || count > Short.MAX_VALUE)
+                throw new IllegalStateException("Maximum lock count exceeded.");
+
+            short newCount = (short) count;
+            long state = ((((long) lockMode) << EXCLUSIVE_SHIFT)) | ((((long) newCount) << COUNT_SHIFT)) | (((long) permit) & PERMIT_MASK);
+
+            // tmp asserts
+            assert permit == getPermitHolder(state);
+            assert count == getCount(state);
+            assert lockMode == getLockMode(state);
+
+            return state;
+        }
+
+        /**
+         * Attempt to acquire the lock in the specified lock mode.
+         * @param permitID - the lock permit object, for exclusive locks, multiple acquires for the same permitID are allowed.
+         *                 This value is a long, but must be <= {@code Integer.MAX_VALUE} and >= {@code Integer.MIN_VALUE}.
+         * @param exclusive - Whether to attempt to acquire the exclusive (true) or shared lock (false).
+         * @return {@code long} < 0 for failure, > 0 for success.
+         */
+        private long internalAcquire(final long permitID, final boolean exclusive) {
+            // constrain permitID to int
+            if (permitID > Integer.MAX_VALUE || permitID < Integer.MIN_VALUE) {
+                throw new IllegalArgumentException("permitID must be between Integer.MIN_VALUE and Integer.MAX_VALUE.");
+            }
+
+            if (isHeldExclusively() && !exclusive) {
+                return -1;
+            }
+
+            int permit = (int) permitID;
+            long state = getState();
+            int count = getCount(state);
+            short mode = getLockMode(state);
+
+            // can't exclusively lock, already in shared mode
+            if (mode == SHARED && exclusive) {
+                return -1;
+            }
+
+            // can't get a shared lock if its already exclusive
+            if (mode == EXCLUSIVE && !exclusive) {
+                return -1;
+            }
```

Comment:
Can L289-291 be removed and just use this block? I think then isHeldExclusively() can be removed as unused.